### PR TITLE
DiffPrivacy ICL Pack - OpenAI Completion LLMs

### PR DESCRIPTION
**Pull request by nerdai**
# Description

- Implementation of "Privacy-Preserving ICL With Differentially Private Few Shot Learning" ([paper](https://openreview.net/pdf?id=oZtt0pRnOl))
- Take in a (new) `LabelledSimpleDataset` to generate differentially private synthetic versions which can be used infinitely many times without taking away to the privacy budget

This is still WIP, but feel free to look around:
1. Support Async calls when generating the `n_split`'s candidate next token probability distributions
2. Support parallel/threaded/async generation of the synthetic examples. I.e., each generation of an example can be handled in parallel.

Fixes # (issue)

## Version Bump?

Did I bump the version in the `pyproject.toml` file of the package I am updating? (Except for the `llama-index-core` package)

- [x] Yes
- [ ] No

## Type of Change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [x] Added new unit/integration tests
- [x] Added new notebook (that tests end-to-end)
- [x] I stared at the code and made sure it makes sense

### CODE REVIEW

File: llama-index-packs/llama-index-packs-diff-private-examples-gen/llama_index/packs/diff_private_simple_dataset/base.py [hunk 1]
```diff
@@ -0,0 +1,294 @@
...
+            )
+
+        if not all(c in sizes for c in self._labels):
+            raise ValueError("Not all labels have sizes.")
+
+        examples = []
+        for label in self._labels:
+            size = sizes[label]
+            for _ in range(size):
+                example = self.generate_dp_synthetic_example(
```
Comment (logan-markewich) [inline llama-index-packs/llama-index-packs-diff-private-examples-gen/llama_index/packs/diff_private_simple_dataset/base.py:285]: if you wanted this to be async, this top-level function is where I would do it. The actually dp generation algorithm seems fairly linear, but if it was using async LLM calls, this top level could be run concurrently ?

File: llama-index-packs/llama-index-packs-diff-private-examples-gen/README.md
Comment (logan-markewich) [inline llama-index-packs/llama-index-packs-diff-private-examples-gen/README.md:1]: Should we take the time to write a better readme? Could be copied from the notebook intro too

Comment (logan-markewich) [reply llama-index-packs/llama-index-packs-diff-private-examples-gen/README.md:1]: lol nvm, the notebook doesn't have an intro

File: llama-index-packs/llama-index-packs-diff-private-examples-gen/notebooks/demo_usage.ipynb
Comment (logan-markewich) [inline llama-index-packs/llama-index-packs-diff-private-examples-gen/notebooks/demo_usage.ipynb:1]: We should probably introduce what's going on here a bit?

File: llama-index-packs/llama-index-packs-diff-private-examples-gen/poetry.lock
Comment (logan-markewich) [inline llama-index-packs/llama-index-packs-diff-private-examples-gen/poetry.lock:1]: no need to commit the lock file

File: llama-index-packs/llama-index-packs-diff-private-examples-gen/README.md
Comment (nerdai) [reply llama-index-packs/llama-index-packs-diff-private-examples-gen/README.md:1]: Yeah, definitely. Will try to tackle all the writing tasks (README and blog post) somewhat together if that makes sense -- hopefully it'll be a bunch of write once, and copy and paste where needed lol.

File: llama-index-packs/llama-index-packs-diff-private-examples-gen/llama_index/packs/diff_private_simple_dataset/base.py (hunk 1 above)
Comment (nerdai) [reply llama-index-packs/llama-index-packs-diff-private-examples-gen/llama_index/packs/diff_private_simple_dataset/base.py:285]: ah yeah so this loop invokes the generate synthetic example as you know.... within this method i think there is room to do more async:

`generate_dp_synthetic_example` does something like this:
 ```
 def generate_dp_synthetic_example():
         synthetic_example = ""
         for _ in max_iterations:
               split_distributions = {}
               for split in splits:
                     # generate candidate next-token (rather their proba distribution) per split
                     next_token_proba_distn = get_next_token_probab_from_llm(split, synthetic_example)
                     
                     split_distributions[split] = next_token_proba_distn
               
               merged_distribution = noisy_aggregation(split_distributions)
               
               next_token = argmax(merged_distribution)
               synthetic_example += next_token
 ```
 
 ☝️ pseudo-code, names of methods differ from actual code just for the sake of demonstration. I think we can do async stuff at the split level as well — that is get the split_distributions via a asyncio.gather()?

File: llama-index-packs/llama-index-packs-diff-private-simple-dataset/llama_index/packs/diff_private_simple_dataset/utils.py
Comment (logan-markewich) [inline llama-index-packs/llama-index-packs-diff-private-simple-dataset/llama_index/packs/diff_private_simple_dataset/utils.py:1]: This file is empty

File: llama-index-packs/llama-index-packs-diff-private-simple-dataset/examples/symptom_2_disease/data/Symptom2Disease.csv
Comment (logan-markewich) [inline llama-index-packs/llama-index-packs-diff-private-simple-dataset/examples/symptom_2_disease/data/Symptom2Disease.csv:1]: Similar to above, it would be nice to not check-in dataset files (if possible)

File: llama-index-packs/llama-index-packs-diff-private-simple-dataset/README.md
Comment (logan-markewich) [inline llama-index-packs/llama-index-packs-diff-private-simple-dataset/README.md:1]: Was this the readme you meant to fill out? or nah?

Comment (nerdai) [reply llama-index-packs/llama-index-packs-diff-private-simple-dataset/README.md:1]: sorry, this was just pushed :)

File: llama-index-packs/llama-index-packs-diff-private-simple-dataset/examples/symptom_2_disease/data/Symptom2Disease.csv
Comment (nerdai) [reply llama-index-packs/llama-index-packs-diff-private-simple-dataset/examples/symptom_2_disease/data/Symptom2Disease.csv:1]: The `LabelledSimpleDataset` could technically be downloaded with `download_llama_dataset` once uploaded there.

But yea, I will remove the .json and .csv files!

File: llama-index-packs/llama-index-packs-diff-private-simple-dataset/examples/basic_demo/agnews.json
[the remark this answers is not in the document]
Comment (nerdai) [reply llama-index-packs/llama-index-packs-diff-private-simple-dataset/examples/basic_demo/agnews.json:1]: Two options:

1. In the notebook, say the user should run `_create_agnews_simple_dataset.ipynb` before running `demo_usage.ipynb`
2. Upload the `agnews.json` as a llama-dataset that can be downloaded from llama-hub like all of our other llama-datasets,

File: llama-index-packs/llama-index-packs-diff-private-simple-dataset/llama_index/packs/diff_private_simple_dataset/utils.py
Comment (nerdai) [reply llama-index-packs/llama-index-packs-diff-private-simple-dataset/llama_index/packs/diff_private_simple_dataset/utils.py:1]: oops! removed! 🙏

File: llama-index-packs/llama-index-packs-diff-private-simple-dataset/examples/basic_demo/agnews.json
[the remark this answers is not in the document]
Comment (nerdai) [reply llama-index-packs/llama-index-packs-diff-private-simple-dataset/examples/basic_demo/agnews.json:1]: going with option 1 - in future we can have these datasets be downloaded from llama hub. I kind of want to force the user to see how to create the `LabelledSimpleDataset` for them to use the pack, versus magically downloading them from hub.

File: llama-index-packs/llama-index-packs-diff-private-simple-dataset/README.md [hunk 2]
```diff
@@ -0,0 +1,125 @@
+# LlamaIndex Packs: `DiffPrivateSimpleDatasetPack`
+
+The `DiffPrivateSimpleDatasetPack` llama pack creates differentially private synthetic
+examples from an original, sensitive dataset.
+
+Differential privacy is a privacy preserving technique that seeks to provide
```
Comment (logan-markewich) [inline llama-index-packs/llama-index-packs-diff-private-simple-dataset/README.md:6]: This is pretty wordy -- can we dumb it down a few notches?

```
Differential Privacy is a privacy preserving technique that obscures source data while preserving original attributes, while minimizing the performance impact on processes that consume the data.
```

?

**Review by logan-markewich — APPROVED** [review]

File: llama-index-packs/llama-index-packs-diff-private-simple-dataset/README.md (hunk 2 above)
Comment (nerdai) [reply llama-index-packs/llama-index-packs-diff-private-simple-dataset/README.md:6]: Yes, we can. I'll take a stab it now :)

Comment (nerdai) [reply llama-index-packs/llama-index-packs-diff-private-simple-dataset/README.md:6]: oh nvm you took a stab at it -- and i like it!
